File: leetcode_learning/code_cpp/141.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */

//使用map记录各个节点出现的次数，如果有一个节点出现两次，则有环
class Solution {
public:
    bool hasCycle(ListNode *head) {
        map<ListNode*, int> node_count;
        while(head!=NULL)
        {
            node_count[head]++;
            if(node_count[head]==2)
                return true;
            head = head->next;
        }
        return false;
    }
};

//快慢指针
public class Solution {
    public boolean hasCycle(ListNode head) {
        if (head == null || head.next == null) {
            return false;
        }
        ListNode slow = head;
        ListNode fast = head.next;
        while (slow != fast) {
            if (fast == null || fast.next == null) {
                return false;
            }
            slow = slow.next;
            fast = fast.next.next;
        }
        return true;
    }
}

File: leetcode_learning/auto-leetcode/141.环形链表/141.py
"""
141.环形链表
难度：简单
链接：https://leetcode.cn/problems/linked-list-cycle/
"""


class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution:
    def hasCycle(self, head: ListNode) -> bool:
        # 解法一：快慢指针，O(1)空间
        slow = head
        fast = head

        while fast and fast.next:
            # 慢指针走一步，快指针走两步
            slow = slow.next
            fast = fast.next.next
            # 快慢指针相遇，说明有环
            if slow is fast:
                return True

        return False

    def hasCycle_hash(self, head: ListNode) -> bool:
        # 解法二：哈希集合，O(n)空间
        seen = set()
        curr = head

        while curr:
            # 节点已在集合中，说明有环
            if curr in seen:
                return True
            seen.add(curr)
            curr = curr.next

        return False


if __name__ == "__main__":
    # 构造有环链表：3 -> 2 -> 0 -> -4 -> 2（环）
    node1 = ListNode(3)
    node2 = ListNode(2)
    node3 = ListNode(0)
    node4 = ListNode(-4)
    node1.next = node2
    node2.next = node3
    node3.next = node4
    node4.next = node2
    assert Solution().hasCycle(node1) == True
    assert Solution().hasCycle_hash(node1) == True

    # 无环链表：1 -> 2 -> None
    node1 = ListNode(1)
    node2 = ListNode(2)
    node1.next = node2
    assert Solution().hasCycle(node1) == False
    assert Solution().hasCycle_hash(node1) == False